### PR TITLE
feat: don't report inbound stream upgrade errors to handler

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

When an inbound stream upgrade fails, there isn't a whole lot we can do about that in the handler. In fact, for several errors, we wouldn't even know which specific handler to target, for example, `NegotiationFailed`. Similiarly, in case of an IO error during the upgrade, we don't know which handler the stream was eventually meant to be for.

## Notes & open questions

~This is a breaking change, thus draft for now.~

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

File: protocols/relay/src/behaviour/handler.rs [hunk 1]
```diff
@@ -474,41 +474,16 @@ impl Handler {
...
-            )) => {
-                self.pending_error = Some(ConnectionHandlerUpgrErr::Upgrade(
-                    upgrade::UpgradeError::Apply(Either::Left(error)),
-                ));
-                return;
-            }
-        };
-
-        self.queued_events.push_back(ConnectionHandlerEvent::Custom(
-            Event::CircuitReqReceiveFailed {
```
Comment (thomaseizinger) [inline protocols/relay/src/behaviour/handler.rs:506]: This event can now be removed/deprecated as well.

File: protocols/relay/src/priv_client/handler.rs [hunk 2]
```diff
@@ -299,41 +299,16 @@ impl Handler {
...
-            )) => {
-                self.pending_error = Some(ConnectionHandlerUpgrErr::Upgrade(
-                    upgrade::UpgradeError::Apply(Either::Left(error)),
-                ));
-                return;
-            }
-        };
-
-        self.queued_events.push_back(ConnectionHandlerEvent::Custom(
-            Event::InboundCircuitReqFailed {
```
Comment (thomaseizinger) [inline protocols/relay/src/priv_client/handler.rs:330]: Same here.

File: protocols/request-response/src/handler.rs [hunk 3]
```diff
@@ -166,30 +166,14 @@ where
...
-                self.pending_events.push_back(Event::InboundTimeout(info))
-            }
-            ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Select(NegotiationError::Failed)) => {
-                // The local peer merely doesn't support the protocol(s) requested.
-                // This is no reason to close the connection, which may
-                // successfully communicate with other protocols already.
-                // An event is reported to permit user code to react to the fact that
-                // the local peer does not support the requested protocol(s).
-                self.pending_events
-                    .push_back(Event::InboundUnsupportedProtocols(info));
```
Comment (thomaseizinger) [inline protocols/request-response/src/handler.rs:180]: Same here.

File: protocols/request-response/src/handler.rs [hunk 4]
```diff
@@ -166,30 +166,14 @@ where
...
         &mut self,
-        ListenUpgradeError { info, error }: ListenUpgradeError<
+        ListenUpgradeError { error, .. }: ListenUpgradeError<
             <Self as ConnectionHandler>::InboundOpenInfo,
             <Self as ConnectionHandler>::InboundProtocol,
         >,
     ) {
-        match error {
-            ConnectionHandlerUpgrErr::Timeout => {
-                self.pending_events.push_back(Event::InboundTimeout(info))
```
Comment (thomaseizinger) [inline protocols/request-response/src/handler.rs:171]: And here.

File: swarm/src/connection.rs [hunk 5]
```diff
@@ -263,12 +263,28 @@ where
...
+                Poll::Ready(Some((
+                    info,
+                    Err(ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Apply(error))),
+                ))) => {
                     handler.on_connection_event(ConnectionEvent::ListenUpgradeError(
                         ListenUpgradeError { info, error },
                     ));
                     continue;
                 }
+                Poll::Ready(Some((_, Err(ConnectionHandlerUpgrErr::Timer)))) => {
```
Comment (thomaseizinger) [inline swarm/src/connection.rs]: While we are at it, remove this `Timer` variant.

File: swarm/src/connection.rs [hunk 6]
```diff
@@ -263,12 +263,28 @@ where
...
+                Poll::Ready(Some((
+                    _,
+                    Err(ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Select(e))),
+                ))) => {
+                    log::debug!("failed to upgrade inbound stream: {e}");
+                    continue;
+                }
+                Poll::Ready(Some((_, Err(ConnectionHandlerUpgrErr::Timeout)))) => {
+                    continue;
+                }
```
Comment (mxinden) [inline swarm/src/connection.rs]: Why not log this error on `debug`?

**Review by mxinden — COMMENTED** [review]
:tada: I am **very** much looking forward to this pull request landing! This has been a thorn in my eye for long.

File: swarm/src/connection.rs (hunk 6 above)
Comment (thomaseizinger) [reply swarm/src/connection.rs]: We could I guess!

**Comment by mxinden** [comment]
@thomaseizinger let me know if you want another review here.

**Comment by thomaseizinger** [comment]
> @thomaseizinger let me know if you want another review here.

Go for it. It is only a draft because it is a breaking change.

File: protocols/dcutr/src/handler/relayed.rs [hunk 7]
```diff
@@ -212,45 +212,12 @@ impl Handler {
...
-                // the remote peer and results in closing the connection.
-                self.pending_error = Some(error.map_upgrade_err(|e| {
-                    e.map_err(|e| match e {
-                        Either::Left(e) => Either::Left(e),
-                        Either::Right(v) => void::unreachable(v),
-                    })
-                }));
-            }
-        }
+        self.pending_error = Some(ConnectionHandlerUpgrErr::Upgrade(UpgradeError::Apply(
```
Comment (mxinden) [inline protocols/dcutr/src/handler/relayed.rs:215]: I think we are on the same page when I say that `libp2p-relay` should not be closing the connection based on `pending_error` via `ConnectionHandlerEvent::Close` further below in `poll`. That said, let's not fix it here but along with https://github.com/libp2p/rust-libp2p/issues/3591.

File: protocols/request-response/CHANGELOG.md [hunk 8]
```diff
@@ -4,6 +4,11 @@
   See [PR 3715].
 - Remove deprecated `RequestResponse` prefixed items. See [PR 3702].
 
+- Remove `InboundFailure::UnsupportedProtocols`.
+  This variant is no longer constructed.
```
Comment (mxinden) [inline protocols/request-response/CHANGELOG.md]:
```suggestion
- Remove `InboundFailure::UnsupportedProtocols` and `InboundFailure::InboundTimeout`.
  This variants is no longer constructed.
```

**Review by mxinden — DISMISSED** [review]
Ready to merge from my end. Great work!

File: protocols/dcutr/src/handler/relayed.rs (hunk 7 above)
Comment (thomaseizinger) [reply protocols/dcutr/src/handler/relayed.rs:215]: Yes, it should be a much easier change once this lands too!

**Review by mxinden — APPROVED** [review]

**Pull request merged**